process/transactionProcessor.go

```diff
@@ -388,22 +400,72 @@ func (tp *TransactionProcessor) getShardByAddress(address string) (uint32, error
 
 // GetTransactionStatus returns the status of a transaction
 func (tp *TransactionProcessor) GetTransactionStatus(txHash string, sender string) (string, error) {
-	if sender != "" {
-		tx, err := tp.getTxWithSenderAddr(txHash, sender, false)
-		if err != nil {
-			return UnknownStatusTx, err
-		}
+	tx, err := tp.getTransaction(txHash, sender, false)
+	if err != nil {
+		return string(data.TxStatusUnknown), err
+	}
+
+	return string(tx.Status), nil
+}
 
-		return string(tx.Status), nil
+func (tp *TransactionProcessor) getTransaction(txHash string, sender string, withResults bool) (*transaction.ApiTransactionResult, error) {
+	if sender != "" {
+		return tp.getTxWithSenderAddr(txHash, sender, withResults)
 	}
 
 	// get status of transaction from random observers
-	tx, err := tp.getTxFromObservers(txHash, requestTypeObservers, false)
+	return tp.getTxFromObservers(txHash, requestTypeObservers, withResults)
+}
+
+// GetProcessedTransactionStatus returns the status of a transaction after local processing
+func (tp *TransactionProcessor) GetProcessedTransactionStatus(txHash string, sender string) (string, error) {
+	tx, err := tp.getTransaction(txHash, sender, true)
 	if err != nil {
-		return UnknownStatusTx, errors.ErrTransactionNotFound
+		return string(data.TxStatusUnknown), err
 	}
 
-	return string(tx.Status), nil
+	return string(tp.computeTransactionStatus(tx, true)), nil
+}
+
+func (tp *TransactionProcessor) computeTransactionStatus(tx *transaction.ApiTransactionResult, withResults bool) transaction.TxStatus {
+	if !withResults {
+		return data.TxStatusUnknown
+	}
+
+	if tx.Status != transaction.TxStatusSuccess {
+		return tx.Status
+	}
+	if findIdentifierInLogs(tx, txFailed) {
+		return transaction.TxStatusFail
+	}
+	containsCompletion := findIdentifierInLogs(tx, txCompleted) || findIdentifierInLogs(tx, scDeploy)
```

Comment:
Unfortunately, a transaction may have multiple `txCompleted` events, and finding the first ones do not indicate actual completion (e.g. think of smart contract calls that perform a few ESDT transfers within the SC execution, one of them intra shard, others cross shard). However, this should be sufficient for now (in my opinion).

Reply:
I think this will be refactored in the upcoming VM v1.5. I would leave it as it is for now, as I do not have a better idea how to implement what you have enumerated :(